### DISCUSSION TITLE
Support for cheap common boards (Mega2560 / STM32F103C8T6)

### DISCUSSION THREAD

**Discussion by TheMinimazer** [EdgeTX/edgetx#3044] [General]
### Is there an existing issue for this feature request?

- [X] I have searched the existing issues

### Is your feature request related to a problem?

I wish to create a custom-shaped transmitter using off-the-shelf components, rather than just buying a pre-built transmitter and pulling it to pieces. OpenTX 2.1 was compatible with the Mega2560 chips and related boards (Arduino 2560 and clones), and it'd be nice to see more easily-accessible open-source firmware availability (and possibly ELRS availability) on cheap hardware. Other options require more know-how than simply buying a dev board and following some relatively simple diagrams and instructions such as what exists for OpenTX 2.1 in [https://github.com/opentx/opentx/tree/2.2/doc/Mega2560](https://github.com/opentx/opentx/tree/2.2/doc/Mega2560) . Most other options do not seem to utilise the JR bay (for external RF modules), which would be useful for customisability of use cases with these cheap boards.

### Describe the solution you'd like

Add support for development boards such as Mega2560 (if possible), STM32F103C8T6 (aka blue pill), STM32F411 (aka black pill), or STM32F407VET6, such that they could be flashed from the companion like a pre-built transmitter.

### Describe alternatives you've considered

1. Deal with lack of recency/features with OpenTX 2.1
2. Compile OpenAVRc builds (I can't easily find a simple companion-style solution in OpenAVRc) and use the many boards for the project rather than just a dev board and pots, switches, etc.
3. Build my own OpenTX 2.3 versions with [https://github.com/jemo07/opentx_STM32F407VET6](https://github.com/jemo07/opentx_STM32F407VET6) and working out pinouts for the board to connect pots, switches, etc.

**Comment by raphaelcoeffic** [comment]
Atmega and F103 won't have enough resources for anything useful. The other 2 are already used in some targets. The question is: what kind of targets should these be? As we are moving towards unified targets, supporting these could be done easily, however it is quite out-of-focus for now.

**Comment by TheMinimazer** [comment]
Not surprised re: AtMega and STM32F103 being weak.

What targets are the other two already used in?

I am not 100% sure on what the question "what kind of targets should these be?" means, but if it is a question of what the use-cases of these boards would be, I am interested in the use of a DIY solution to obtaining a land/surface (a.k.a. pistol grip) transmitter  for controlling complex (4+ channels) land vehicles. I guess there could also be use for these in creating more accessible/disability-friendly controllers, but that would not be what I am intending.

It is a shame there are no commercially-available surface transmitters with OpenTX/EdgeTX, and it'd be nice if one could be created easily (just plug in and upload firmware, rather than having to read code to find pinouts and then compile the code by oneself) with the ability to swap external modules (e.g. swapping between ELRS and 4in1 modules).

**Comment by rotorman** [comment]
> What targets are the other two already used in?

Here a list of some microcontrollers known to me by the EdgeTX supported radios presently:
STM32F205VGT6
STM32F205ZET6
STM32F407VET6
STM32F429BIT6
STM32F439BIT6
First 3 are for black and white screen radios, the latter 2 are for color screen radios with beefier graphics and external SDRAM.

**Comment by TheMinimazer** [comment]
I found that the STM32F407VET6 is used in the RadioMaster Tx12. So if that microcontroller is supported already, does that mean that I could:
1. Buy a [STM32F407VET6 development board with screen from AliExpress](https://www.aliexpress.com/item/4000953664266.html)
2. Use Zadig to make it bootloadable as per Page 13 of the [RadioMaster Tx12 Manual](https://manuals.plus/wp-content/sideloads/radiomaster-tx12-manual-optimized.pdf)
3. Flash it with the EdgeTx Companion, by selecting 'Radio Type' as 'Radiomaster Tx12'
4. Find the correct pin assignments for what components connect to which pins in one of the .h files in [EdgeTX/edgetx/tree/main/radio/src/targets/common/arm/stm32](https://github.com/EdgeTX/edgetx/tree/main/radio/src/targets/common/arm/stm32)
5. Build a DIY transmitter around it by using switches, potentiometers, etc. connected to the dev board

Would that work? And if so, which file should I be looking in for the pin assignments?

**Comment by rotorman** [comment]
You could try to use a STM32F407VET6 Development Board, but would need to likely adjust the I/O pins in the code (assuming the external oscillator frequency matches, otherwise would need to adjust also the PLL values).
Present EdgeTX code for black and white targets with F407, can drive only 128x64 pixel or 212x64 pixel monochrome displays, not the color display included in that kit.
Instead of Zadig, best have a look at official ST tool - STM32CubeProgrammer https://www.st.com/en/development-tools/stm32cubeprog.html
Also consider getting yourself a Serial Wire Debug (SWD) capable in-circuit-debugger, such as ST-Link https://www.st.com/en/development-tools/st-link-v2.html or J-Link https://www.segger.com/products/debug-probes/j-link/models/j-link-edu-mini/ or similar. As, if you need to modify the code to get your changed I/O configuration running, it is easier to debug the changes via in-circuit-debugging.

For the I/O mapping of switches have a look at: https://github.com/EdgeTX/edgetx/blob/main/radio/src/targets/taranis/hal.h and https://github.com/EdgeTX/edgetx/blob/main/radio/src/targets/taranis/board.h

**Comment by gagarinlg** [comment]
If you wait a bit, all the I/O config will get much easier, when the switch-refactor branch is merged

**Comment by microbearsmb** [comment]
I'm also interested in this possibility; has that branch been merged yet and if so, what parts of this undertaking are made easier?

**Reply by pfeerick** [reply]
Not yet - the PR in question is https://github.com/EdgeTX/edgetx/pull/2631, and will be part of 2.10. It will make mapping of switches to IOs simpler as you won't need to alter it in as many places in the codebase... ideally just a couple of configuration files.

**Comment by stefanfoulis** [comment]
@TheMinimazer Did you end up finding suitable development board and a process to get EdgeTX running on it?
I'm curious, because I'd also like to build a PIG style transmitter.

**Reply by TheMinimazer** [reply]
I have not looked into this any further, as I am not very knowledgeable in regards to programming for microprocessors. As such, I am waiting for version 2.10, which as @pfeerick mentioned will make IO mapping easier due to config file manipulation rather than changing the code directly.

If you feel up to it, you could have a go at following @rotorman 's 12-01-2023 reply to this thread:
[https://github.com/EdgeTX/edgetx/discussions/3044#discussioncomment-4689082](https://github.com/EdgeTX/edgetx/discussions/3044#discussioncomment-4689082)

**Comment by DAMskill** [comment]
I'm going to build my own Radio soon.
Don't want to spend a few hundred euro on a radio,
that eventually does not do the exact thing I need it to do.

It will be the first time in my life compiling code
Can someone please explain the steps involved, to build your own version of EdgeTX?

Maby a stipid question. 
What are the steps involved work on a Raspberry Pi for example?

Here in this comment I explain a bit more of my plan. 
Cool project by the way, This genius got EdgeTx working on a ESP32-S3.
https://github.com/JunOllyLi/espidf5_edgetx/issues/1#issuecomment-2094883167

To keep things simple at first. I would like to give this board a try, since it will have its first support in EdgeTx 3.x.:
https://www.st.com/en/evaluation-tools/stm32-nucleo-boards.html

=========
Why not team-up to get this plan figured-out.
Lets build the best Modular EdgeTX Radio ever,  

The DIY Till I Die board :)
Who is in for some teamwork?
=========

**Reply by pfeerick** [reply]
> What are the steps involved work on a Raspberry Pi for example?

Probably not going to happen... the Raspberry Pi processor is simply not suited to running a radio - it is geared towards general purpose processing, not real-time/time sensitive operations. 

If, however, you actually mean the Raspberry Pi Pico, that would be interesting to see in the future, and like with the ESP32 port could happen.
